physical/postgresql/transaction.go

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2024 OpenBao a Series of LF Projects, LLC
+// SPDX-License-Identifier: MPL-2.0
+
+package postgresql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"sync"
+
+	"github.com/openbao/openbao/sdk/v2/physical"
+)
+
+type PostgreSQLBackendTransaction struct {
+	l  sync.Mutex
+	b  *PostgreSQLBackend
+	tx *sql.Tx
+
+	readOnly       bool
+	haveFinishedTx bool
+}
+
+func (b *PostgreSQLBackend) newTransaction(ctx context.Context, readOnly bool) (physical.Transaction, error) {
+	// Grab a transaction permit pool entry so that we can limit the number of
+	// concurrent transactions.
+	b.txnPermitPool.Acquire()
+
+	tx, err := b.client.BeginTx(ctx, &sql.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
```

Comment:
The only real surprise was needing to use `LevelRepeatableRead` isolation: I'd get errors while reading entries inside my read-only transactions, that were deleted by another write transaction if I used the default `LevelReadCommitted`.